src/Lean/Syntax.lean

```diff
@@ -530,4 +530,27 @@ def Stack.matches (stack : Syntax.Stack) (pattern : List $ Option SyntaxNodeKind
     |>.all id)
 
 end Syntax
+
+namespace TSyntax
+
+/--
+Applies the given function to every subsyntax.
+
+Like `Syntax.replaceM` but for typed syntax.
+(Note there are no guarantees of type correctness here.)
+-/
+def replaceM [Monad m] (f : Syntax → m (Option Syntax)) (stx : TSyntax k) : m (TSyntax k) :=
+  .mk <$> stx.1.replaceM f
+
+end TSyntax
+
+/--
+Constructs a typed separated array from elements.
+The given array does not include the separators.
+
+Like `Syntax.SepArray.ofElems` but for typed syntax.
+-/
+def Syntax.TSepArray.ofElems {sep} (elems : Array (TSyntax k)) : TSepArray k sep :=
```

Comment:
We should move this in front of https://github.com/leanprover/lean4/blob/5cc0c3f145ffa623a617ca31387d7f589802c1d3/src/Init/Meta.lean#L566-L567 and make use of it there

Reply:
Done.